### ISSUE TITLE
aur-build should not require -f for VCS packages

### ISSUE THREAD

**Issue by lahwaacz** [AladW/aurutils#500]
From `aur-build(1)` in aurutils-2.0.1-1:

>       -f, --force
>              Continue the build process if a package with the same name is
>              found (ignoring the extension). Existing packages will be moved
>              with a ~ suffix.

Unfortunately the check if a package with the same name is found seems to be done before the `pkgver()` function of VCS packages is executed, which is very un-makepkg-like.

**Comment by FichteFoll** [comment]
The current recommended method to update VCS packages is to use the `--rebuild` flag.

**Comment by lahwaacz** [comment]
@FichteFoll `aur-build` does not have a `--rebuild` flag.

**Comment by AladW** [comment]
How about adding an `--update-pkgver` option to `aur-build` which runs `makepkg -od --skipinteg`?

**Comment by lahwaacz** [comment]
I'm not acquainted with the exact command sequence in `aur-build`, but `-d` is probably not a good idea because `pkgver()` usually requires some makedepends (e.g. `git`).

**Comment by AladW** [comment]
The issue is that -d installs _all_ makedepends, not just those required to update pkgver. If you used e.g. a chroot build, that would be suboptimal.

I suppose you could check for E_MISSING_MAKEPKG_DEPS, but it also does not distinguish between git/hg/.. and regular makedepends. (The only distinction is in the output on stderr.)

On Fri, 11 Jan 2019 14:14:59 -0800
Jakub Klinkovský <notifications@github.com> wrote:

> I'm not acquainted with the exact command sequence in `aur-build`, but `-d` is probably not a good idea because `pkgver()` usually requires some makedepends (e.g. `git`).
> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/AladW/aurutils/issues/500#issuecomment-453674287


-- 
Alad Wenter <alad@archlinux.org>

**Comment by lahwaacz** [comment]
That's an issue with the build system, not with aurutils. Installing all dependencies would still be faster than rebuilding the whole package, which is what `-f` does, even if pkgver did not change. Also, at least on my system, the installation of dependencies in chroot is _very fast_ due to an SSD and 1 Gbit internet connection (but packages can also be cached, which can be expected for VCS), unless it contains some huge package like cuda.

**Comment by AladW** [comment]
Well, I don't have a 1 Gbit connection... some proposals:

1. I guess the most "makepkg-like" way would be to error when git/bzr/etc. are not found. Example:

user@foxden ~/aurutils/makepkg % ../lib/aur-build --update-pkgver
==> ERROR: build: makepkg exit code: 1
==> Making package: aurutils-git 2.0.1-1 (Sat Jan 12 12:08:39 2019)
==> WARNING: Skipping dependency checks.
==> Retrieving sources...
/usr/share/makepkg/source/git.sh: line 54: git: command not found
==> ERROR: /home/user/aurutils/makepkg/aurutils is not a clone of https://github.com/AladW/aurutils
    Aborting...

2. I've found some way to only install the VCS packages, but that requires duplicating some of makepkg's internals.

declare -A missing_deps

( PATH= . "$tmp"/makepkg.conf
  PATH= . PKGBUILD

  for i in "${source[@]}"; do
      proto=$(get_protocol "$i")
      proto=${proto%%+*}

      for handler in "${VCSCLIENTS[@]}"; do
          if [[ ${handler%%::*} == $proto ]]; then
              client=${handler##*::}
              break
          fi
      done

      if [[ $client ]]; then
          pacman -T "$client" || missing_deps[$client]=1 ;;
      else
          error "$argv0: Unknown download protocol: $client"
          exit 1
      fi
  done

  # automatically install missing dependencies
  sudo pacman -S --asdeps "${!missing_deps[@]}"
)

3. Instead of choosing on -o or -od, the user could specify the "prepare" command. I guess that's similar to bauerbill's "pre-build" functionality.

On Fri, 11 Jan 2019 23:41:08 -0800
Jakub Klinkovský <notifications@github.com> wrote:

> That's an issue with the build system, not with aurutils. Installing all dependencies would still be faster than rebuilding the whole package, which is what `-f` does, even if pkgver did not change. Also, at least on my system, the installation of dependencies in chroot is _very fast_ due to an SSD and 1 Gbit internet connection (but packages can also be cached, which can be expected for VCS), unless it contains some huge package like cuda.
> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/AladW/aurutils/issues/500#issuecomment-453727571


-- 
Alad Wenter <alad@archlinux.org>

**Comment by AladW** [comment]
Note that 1. would suffice if you have devtools installed, since it depends on all common VCS clients.

**Comment by lahwaacz** [comment]
I have devtools (required for aur-chroot anyway) so I'm fine with 1.

**Comment by AladW** [comment]
633cb12f38b8753a18758f7b902df8f7b2e2a527

Thanks for the report!

On Sat, 12 Jan 2019 05:44:23 -0800
Jakub Klinkovský <notifications@github.com> wrote:

> I have devtools (required for aur-chroot anyway) so I'm fine with 1.
> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/AladW/aurutils/issues/500#issuecomment-453748681


-- 
Alad Wenter <alad@archlinux.org>

**Comment by AladW** [comment]
Note that I've renamed --run-pkgver to --pkgver, as will be reflected in
the next version (probably 2.2.0).

**Comment by FichteFoll** [comment]
Would it be reasonable to add this option to AUR-sync too?

**Comment by AladW** [comment]
I wonder if we still need `aur-vercmp-devel` at all in that case?

**Comment by maximbaz** [comment]
Ah well, unless I'm missing something in that case it seems there aren't that many benefits to adding `--pkgver` to `aur-sync`...

**Comment by AladW** [comment]
Right, `--pkgver` is basically a special case of `aur-srcver`. Though there's no downside in supporting the option in `aur-sync` either. Maybe @FichteFoll can clarify.

**Comment by FichteFoll** [comment]
Using `aur-vercmp-devel` first to check out the updated packages and then submitting those to `aur-sync --nover-shallow` is the explicit operation and there isn't anything particularly wrong with that. I attempted to use `--pkgver` with `aur-sync` recently but was set back for a moment when I realized it was added to `aur-build` (which makes more sense, of course).

Adding `--pkgver` to `aur-sync` would allow requesting an update of a VCS package as needed but without having to update the entire repo. The only reason I do that is because I have a couple VCS packages in my aur repo that take quite a bit of time to build and don't need regular updates for, however. Might just be a special case?
